r1cs/gadgets/std/src/multieq.rs

```diff
@@ -0,0 +1,127 @@
+use algebra::{PrimeField, FpParameters};
```

Comment:
Please not that you are not using at all the optimizations within this ConstraintSystem.
You pass it as a type bound for the sha256 gadget, but the actual place where it's used is the UInt32::add_many function you left untouched. (See the diff between our addmany (left-side) vs the zcash one )

https://www.diffchecker.com/pFPqK7Hd

That's why you are not able to achieve Bellman's 25840 constraints vs the 25996 that we have.
The difference is not huge (0.3 constraints per bit) but since we already have everything implemented I propose to use the ZCash optimization.
Please replace our UInt32::addmany() function with Bellman's one (make sure to use the MultiEq CS and replace all the instances in GingerLib where we use the addmany to pass the MultiEq CS (not many of them, just a bunch of tests))